platform-sdk/docs/proposals/DP-0x/0X - Tss block signing - Cryptography APIs.md

```diff
@@ -0,0 +1,695 @@
+`t` (t=threshold) valid signatures are aggregated. If the threshold is met and the signature is valid, the library will respond with true; if not, it will be false.
+
+
+#### Summary diagram
+The following diagram exposes the steps necessary for bootstrapping and using the library. In orange are all the steps where the library is involved.
+The rest of the steps need to happen but are outside the scope of the described library.
+
+![img_12.png](img_12.png)
+
+
+### Architecture
+To implement the functionality detailed in the previous section, the following code structure is proposed:
+
+![img_10.png](img_10.png)
+
+1. **TSS Lib**: The consensus node will use the TSS library to create shares, create TSS messages to send to other nodes,
+   assemble shared public keys (ledgerId), and sign the block node Merkle tree hash.
+2. **Bilinear Pairings Signature Library**: This library provides cryptographic objects (PrivateKey, PublicKey, and Signature)
+   and operations for the block node and consensus node to sign and verify signatures. The consensus node uses this library indirectly through the TSS Library.
+3. **Bilinear Pairings API**: Layer of abstraction to be included at compilation time providing the cryptography primitives
+   and the arithmetic operations for working with a specific EC curve and the underlying Groups, Fields, and Pairings
+   and minimizes the impact of changing or including different curve implementations.
+4. **Bilinear Pairings Impl**: This is an implementation of the Bilinear Pairings API that will be loaded at runtime using Java’s SPI.
+   Multiple implementations can be provided to support different types of curves. It allows changing or switching dependencies.
+5. **Native Support Lib**: Provides a set of generic functions loading native libraries in different system architectures when packaged in a jar
+   using a predefined organization so they can be accessed with JNI.
+6. **Arkworks[https://github.com/arkworks-rs]**:  arkworks is a Rust ecosystem for zkSNARK programming.
+   Our implementation uses it as the library responsible for elliptic curve cryptography.
+7. **EC-Key Utils ** is a utility module that enables the node operator to generate a bootstrapping public/private key pair.
+
+### Module organization and repositories
+![img_13.png](img_13.png)
+1. **hedera-cryptography**: This is a separate repository for hosting cryptography-related libraries.
+   It is necessary to facilitate our build process, which includes Rust libraries. It also provides independent release cycles between consensus node code and block node code.
+2. **swirlds-native-support**: Gradle module that enables loading into memory compiled native libraries so they can be used with JNI.
+3. **swirlds-cryptography-tss**: Gradle module for the TSS Library.
+   This library-only client is the consensus node so that it will remain close to it in the `hedera-services` repository under the `platform-sdk` folder.
```

Comment:
Consider the following rewording:

"This library-only client is the consensus node so that it will remain close to it in the `hedera-services` repository"... ->"This library's only client is the consensus node, so it will be in the `hedera-services` repository"...